File: Data Structures & Algorithms/reorder-linked-list/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    // Redo the solution without any helping methods/functions
    void reorderList(ListNode* head) {
        // A list with 0 or 1 nodes can't be reordered
        if (head == nullptr || head->next == nullptr)
            return;

        // First compute the midpoint of the list using a pair of slow and fast pointers
        ListNode* slow = head; // Node 1 (1-based index)
        ListNode* fast = head->next; // Node 2 (1-based index)
        // Advance both until fast is either the tail (even nodes), or nullptr (odd nodes)
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        // Now if there were i iterations, then slow = 1 + i, fast = 2 + 2i
        // We'll take slow to be the tail of the first half list
        // So slow->next is the head of the second half list
        ListNode* second = slow->next;
        slow->next = nullptr;

        // Reverse the second half
        ListNode* prev = nullptr;
        while (second != nullptr) {
            ListNode* second_next = second->next;
            second->next = prev;
            prev = second;
            second = second_next;
        }
        // Now second == nullptr, so prev is the head of the reversed list

        // Merge the lists with heads first and prev (head of reversed second half, renamed second). 
        // The second list is equal or shorter than the first, so we only need to check
        ListNode* first = head;
        second = prev;
        while (second != nullptr) {
            ListNode* first_next = first->next;
            ListNode* second_next = second->next;
            first->next = second;
            second->next = first_next;
            first = first_next;
            second = second_next;
        }
        // Now second == nullptr, but it's possible first != nullptr, 
        // in which case it's the last node of the merged list
    }
};

File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Observe that the original list has HEAD=head
        #   head -> node1 -> node2 -> ... -> tail -> None
        # We want the reversed list
        #   None <- head <- node1 <- node2 <- ... <- tail which has HEAD=tail
        
        # A list with 0 or 1 elements is the same reversed
        if not head or not head.next:
            return head

        # List has 2 or more elements
        cur = head
        ahead = cur.next
        while ahead:
            twoAhead = ahead.next
            ahead.next = cur
            cur = ahead
            ahead = twoAhead

        # Now ahead == None, so cur is the tail of the original list, or the head of the new list
        # The original head is now the tail, so set it's 'next' to None
        head.next = None
        return cur